File: api-delpi/app/domain/entities/lmp/lmp.py
# app/domain/entities/lmp/lmp.py
from dataclasses import dataclass, field, asdict
from typing import Optional, List

from app.shared.utils.spreadsheet_date import format_date_iso

from .lmp_product import LMPProduct
from .lmp_history_event import LMPHistoryEvent


@dataclass
class LMP:
    branch: Optional[str] = None
    sale_number: str = ""
    sale_description: str = ""
    listing_kind: Optional[str] = None

    # Engineering summary
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    reference_revision: Optional[str] = None
    measurement_revision: Optional[str] = None
    engineering_status: Optional[str] = None
    qtd_engineering_entries: Optional[int] = None
    qtd_engineering_closed: Optional[int] = None
    qtd_advanced_from_engineering: Optional[int] = None
    qtd_returned_from_engineering: Optional[int] = None
    engineering_total_minutes: Optional[int] = None

    # PI summary
    qtd_pi: Optional[int] = None

    # Costumer
    costumer_code: Optional[str] = None
    costumer_store: Optional[str] = None
    costumer_name: Optional[str] = None

    # Seller
    seller_code: Optional[str] = None
    seller_name: Optional[str] = None

    # Products
    list_products: Optional[List[LMPProduct]] = field(default_factory=list)

    # Engineering / workflow history (AIJ010)
    list_history: Optional[List[LMPHistoryEvent]] = field(default_factory=list)

    def to_dict(self) -> dict:
        data = asdict(self)
        data["start_date"] = format_date_iso(data.get("start_date"))
        data["end_date"] = format_date_iso(data.get("end_date"))
        history = data.get("list_history") or []
        if history:
            data["list_history"] = [
                {
                    **event,
                    "start_date": format_date_iso(event.get("start_date")),
                    "limit_date": format_date_iso(event.get("limit_date")),
                    "end_date": format_date_iso(event.get("end_date")),
                    "next_start_date": format_date_iso(event.get("next_start_date")),
                }
                if isinstance(event, dict)
                else event
                for event in history
            ]
        return data

File: plugins/dashboard-lmps/src/types/lmp.ts
export type LmpStatus = "Pontual" | "Atrasado" | "Andamento" | "Retornada";
export type LmpNivel = "Nível 1" | "Nível 2" | "Nível 3";

export type LmpProduct = {
  code: string;
  description: string;
  group_code: string;
  type: string;
  qtd_pi?: number | null;
};

export type LmpHistoryEvent = {
  revision: string;
  process_code: string;
  stage_code: string;
  process_label?: string | null;
  stage_label?: string | null;
  start_date?: string | null;
  start_time?: string | null;
  limit_date?: string | null;
  limit_time?: string | null;
  end_date?: string | null;
  end_time?: string | null;
  duration_minutes?: number | null;
  duration_display?: string | null;
  status?: string | null;
  status_label?: string | null;
  history_flag?: string | null;
  is_engineering?: boolean;
  is_engineering_flow?: boolean;
  is_open?: boolean;
  is_late?: boolean;
  is_current?: boolean;
  is_engineering_entry?: boolean;
  flow_transition?: string | null;
  flow_transition_label?: string | null;
  flow_transitions?: string[];
  flow_transition_labels?: string[];
};

export type LmpListingKind = "LMP" | "AMOSTRA" | "OUTRO";

export type LmpItem = {
  branch?: string | null;
  sale_number: string;
  sale_description: string;
  listing_kind?: LmpListingKind | null;
  start_date?: string | null;
  end_date?: string | null;
  reference_revision?: string | null;
  measurement_revision?: string | null;
  homolog_revision?: string | null;
  homolog_date?: string | null;
  cycle_index?: number | null;
  engineering_status?: string | null;
  qtd_engineering_entries?: number;
  qtd_engineering_closed?: number;
  qtd_advanced_from_engineering?: number;
  qtd_returned_from_engineering?: number;
  engineering_total_minutes?: number;
  qtd_pi?: number | null;

  costumer_code?: string | null;
  costumer_store?: string | null;
  costumer_name?: string | null;
  seller_code?: string | null;
  seller_name?: string | null;
  list_products?: LmpProduct[];
  list_history?: LmpHistoryEvent[];
};

export type LmpDashboardItem = LmpItem & {
  nivel: LmpNivel;
  dias_uteis_sla: number;
  data_limite?: string | null;
  lead_time_util?: number | null;
  status: LmpStatus;
};

export type LmpDetailData = LmpDashboardItem & {
  sla_minutos?: number | null;
};

export type LmpsEvolutionDatum = {
  periodo: string;
  mediaLead: number;
  propostas: number;
};

export type Page<T> = {
  items: T[];
  total: number;
  page: number;
  page_size: number;
};

export type ListLmpsParams = {
  start_date?: string;
  end_date?: string;
  branch?: string;
  listing_type?: string;
  page?: number;
  page_size?: number;
};

export type ApiDelpiResponseMeta = {
  dataVersion?: string;
  operationId?: string;
  entity?: string;
  shape?: string;
  pagination?: Record<string, unknown>;
  fields?: Record<string, string>;
  relatedRoutes?: Record<string, string>;
  sections?: Array<Record<string, unknown>>;
};

export type ApiDelpiErrorPayload = {
  code?: string;
  recoverable?: boolean;
};

export type ApiSuccessResponse<T> = {
  success: boolean;
  message?: string;
  data: T;
  meta?: ApiDelpiResponseMeta;
  error?: ApiDelpiErrorPayload | null;
};

export function unwrapApiDelpiEnvelope<T>(
  response: ApiSuccessResponse<T>,
  fallbackMessage: string,
): T {
  if (response.success === false) {
    throw new Error(response.message?.trim() || fallbackMessage);
  }
  return response.data;
}